### PR TITLE
Support context parallelism for Qwen3.5

### PR DESCRIPTION
**Pull request by haok1402**
## Why

Qwen3.5 is hybrid, three GatedDeltaNet layers per softmax-attention layer, and the two need the sequence sharded differently. Softmax attention needs zigzag, where rank `r` holds blocks `r` and `2 * cp_size - r - 1` so the causal triangle spreads evenly instead of leaving rank 0 idle. A recurrence cannot use that: its state flows from rank `r - 1` into rank `r`, so it needs rank order to be sequence order, which is contiguous, blocks `2r` and `2r + 1`. The conflict is structural, since balancing causal work means pairing an early block with a late one, which is exactly what destroys the ordering a recurrence needs. `Qwen35MoeModel` previously raised `NotImplementedError` when `cp_size > 1`.

## What changes

A new module, `pithtrain/operators/cp_sequence.py`, owns both layouts. `zigzag_spans` is the single definition of the zigzag partition, which the loader turns into memmap reads and `forward_posemb` into position ids; a disagreement between those two would mistrain silently rather than raise, so there is one definition. `zigzag_to_contiguous` and `contiguous_to_zigzag` convert the hidden stream in a single all-to-all, because both layouts cut the sequence into the same `2 * cp_size` blocks and give each rank two of them, differing only in which two, making the conversion an exchange rather than a gather. `prepend_conv_state` left-pads a shard with the `kernel_size - 1` tokens before it, fetched from the previous rank in one P2P hop.

Conversions are hoisted to the ends of each run of linear-attention layers, 2 all-to-alls per run instead of 6 at 3:1. Each layer derives its flags from the global `config.layer_types` at construction, so both sides of a pipeline boundary agree without negotiating and P2P may carry either layout. The conversion back to zigzag happens on the way out of a run rather than on entry to the next layer, which keeps every run self-closing: the epilog, the loss and the loader never learn a second layout exists, and correctness does not depend on where `layer_types` ends. `gated_delta_rule` gains an optional `cp_group` that routes to FLA's context-parallel path, and softmax layers switch to the existing `ring_attention_func`, which already expects zigzag.

Packed sequences under context parallelism are not implemented. Qwen3 and DeepSeek-V2 assert `cu_seqlens is None or cp_size == 1`; Qwen3.5 is pretraining-only and already asserted `cu_seqlens is None`.

Also here: `tests/utilities.py` sized the test mesh as `pp * cp * ep`, the pre-folding requirement #93 removed, and now uses `pp * lcm(cp, ep)`, which is what makes `cp=3` testable on 3 GPUs. And `benchmarks/pretraining/qwen3.5-122b-a10b/` adds a 48-layer scenario at `pp4-dp2-cp4-ep8-seq8192` on 4x8 H200.

## Correctness validation

| check | result |
|---|---|
| `test_dualpipev` qwen3.5, `cp=4 seq=1024 pp=2 ep=2`, against the single-device reference | loss and gradients match |
| `test_dualpipev` qwen3.5 at `cp` 1, 2 and 4 | pass |
| `tests/operators/test_cp_sequence.py`, `cp` in {2, 3, 4, 8} | 12 passed |

`tests/test_dualpipev.py` forces `pp = ep = cp = 1` and `cp_rank = 0` before building the reference, so it runs the whole sequence on one device with dense flash attention, no relayout, no conv shift, no ring attention and no FLA CP path, then compares the loss and the parameter gradients against the sharded run. Gradients are compared by cosine difference against a 1e-2 threshold, skipping parameters whose reference gradient sits at the bf16 noise floor. There is no base-vs-feature comparison because the base branch raises `NotImplementedError` when constructing a Qwen3.5 model at `cp_size > 1`, so no mesh that exercises this change runs on both branches. Both shipped configs end in `full_attention`, so the conversion back to zigzag at a model boundary is not exercised.

## Performance

qwen3.5-35b-a3b on 4x8 H200, `pp=4 ep=8`. Doubling `cp` doubles the sequence, so every row holds 8192 tokens per rank, 32 micro-batches per rank and 2,097,152 tokens per optimizer step, and differs only in global context. Throughput is the median over steps 10 to 25; per-step spread is 0.6% to 1.3%. All rows set `training.benchmark = True`, which force-balances MoE routing, so these are idealized-routing figures and carry no loss signal.

| cp | dp | gbs | sequence | tokens/s | peak alloc | MFU |
|---|---|---|---|---|---|---|
| 1 | 8 | 256 | 8,192 | 294,700 | 90.95 GiB | 18.9% |
| 2 | 4 | 128 | 16,384 | 258,300 | 91.23 GiB | 18.2% |
| 4 | 2 | 64 | 32,768 | 224,100 | 91.24 GiB | 18.6% |
| 8 | 1 | 32 | 65,536 | 178,000 | 91.25 GiB | 19.3% |

Peak allocated memory is flat to 0.33% across an 8x increase in context. Parameter and optimizer memory are identical in every row by construction, since FSDP shards attention parameters over the flattened `(dp, cp)`, which is the 8-rank pipeline stage regardless of `cp`, and expert parameters over `expt_dp = stage / ep = 1`.

Throughput falls 40% from `cp=1` to `cp=8` because FLOPs per token rise 70%: causal attention is quadratic in the global sequence, so its share of model FLOPs goes from 9.9% at 8K to 46.9% at 64K. MFU stays within 1.1 points across the sweep.

`cp=1` above 8K was not run, so no claim is made about where it stops fitting. Note also that peak allocated is not device occupancy: `nvidia-smi` showed about 121 GiB against the 91 GiB reported here, the difference being the caching allocator's reserved pool, the CUDA context and NCCL buffers.

MFU is model FLOPs against 989 TFLOP/s, dense BF16 on H200 SXM, not the 1,979 sparsity figure. `training.fp8` is false in these runs, and nothing in the repo recomputes activations.

```
MFU = tokens_per_second * F(S) / (32 * 989e12)

F(S) = 18.2398e9 + 245,760 * S      FLOPs per token, forward + backward

  = 6 * [ 10*h*(2*n_q*d + 2*n_kv*d) + 10*n_q*d*h        attn proj (q+gate, k, v, o)  1.636 G
        + 30*h*(2*q_dim + 2*v_dim + 2*H_v) + 30*v_dim*h  GDN in_proj + out_proj       6.063 G
        + 30*k_c*(2*q_dim + v_dim)                       GDN depthwise conv1d         0.006 G
        + 40*3*h*f*K + 40*3*h*f_s                        MoE top-8 + shared           6.795 G
        + 40*(h*E + h)                                   router + shared gate         0.126 G
        + h*V ]                                          lm_head, untied              3.051 G
    + 2*30*H_v*35.75*C*D                                 gated delta rule core        0.562 G
    + 6*10*n_q*d*S                                       causal attention core    245,760*S

h=2048 V=248320 n_q=16 n_kv=2 d=256 q_dim=2048 v_dim=4096 H_v=32 D=128 C=64 k_c=4 E=256 K=8 f=f_s=512
```

### CODE REVIEW

**Comment by haok1402** [comment]
@claude review
